system-addon/lib/PersistentCache.jsm

```diff
@@ -0,0 +1,92 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const {utils: Cu} = Components;
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+XPCOMUtils.defineLazyModuleGetter(this, "OS", "resource://gre/modules/osfile.jsm");
+XPCOMUtils.defineLazyGetter(this, "gTextDecoder", () => new TextDecoder());
+
+// Map of filenames to the latest data in them.
+XPCOMUtils.defineLazyGetter(this, "gInMemoryCache", () => new Map());
+// A list of cache files that has already been loaded.
+XPCOMUtils.defineLazyGetter(this, "gFilesLoaded", () => []);
```

Comment:
Hmmmm......... why do we want to allow for caches that aren't loaded? It seems that these two could be combined into one to avoid some odd condition, e.g., setting then getting resulting in mismatch between what's on disk and in memory.

Maybe just have a `const gCache = new Map()` that maps from the name to a promise of an object.

Reply:
I was trying to handle `set` being called before load was done, but I see how you handled that nicely in your proposal. I like it. You're right, there currently isn't a need for a shared global. I was thinking about loading the cache from `ActivityStream.jsm` which would possibly require it or having a shared cache instance, but I'm not sure that would be needed anyway.

I'll work through these changes, I think the end result will be nicer 👍